help_function/src/onedpl_test_tagged_pointer.cpp

```diff
@@ -0,0 +1,153 @@
+//
+//
+// ===----------------------------------------------------------------------===//
+
+// Adapted from onedpl_test_device_ptr.cpp
+
+#include <oneapi/dpl/algorithm>
+#include <oneapi/dpl/execution>
+#include <oneapi/dpl/iterator>
+#include <oneapi/dpl/numeric>
+
+#include "dpct/dpct.hpp"
+#include "dpct/dpl_utils.hpp"
+
+#include <sycl/sycl.hpp>
+
+#include <iostream>
+
+// Used to test operator-> behavior
+struct integer_wrapper {
+  int val;
+};
+
+template <typename String, typename _T1, typename _T2>
+int ASSERT_EQUAL(String msg, _T1 &&X, _T2 &&Y) {
+  if (X != Y) {
+    std::cout << "FAIL: " << msg << " - (" << X << "," << Y << ")" << std::endl;
+    return 1;
+  } else {
+    std::cout << "PASS: " << msg << std::endl;
+    return 0;
+  }
+}
+
+template <typename SystemTag> int test_tagged_pointer_manipulation(void) {
+  int failing_tests = 0;
+  constexpr ::std::size_t n = 20;
+  SystemTag system;
+  ::std::string sys = ::std::is_same_v<dpct::host_sys_tag, SystemTag>
+                          ? "dpct::host_sys_tag"
+                          : "dpct::device_sys_tag";
+  std::string int_ptr_name = "dpct::tagged_pointer<" + sys + ", int>";
+  std::string int_wrapper_name =
+      "dpct::tagged_pointer<" + sys + ", integer_wrapper>";
+  std::string void_ptr_name = "dpct::tagged_pointer<" + sys + ", void>";
+
+  dpct::tagged_pointer<SystemTag, void> void_ptr_beg =
+      dpct::malloc(system, sizeof(int) * n);
+  dpct::tagged_pointer<SystemTag, int> int_ptr_beg =
+      static_cast<dpct::tagged_pointer<SystemTag, int>>(void_ptr_beg);
+
+  dpct::tagged_pointer<SystemTag, void> void_ptr_beg2 =
+      static_cast<dpct::tagged_pointer<SystemTag, void>>(int_ptr_beg);
+  failing_tests += ASSERT_EQUAL(void_ptr_name + " conversion operator",
+                                void_ptr_beg == void_ptr_beg2, true);
+
+  dpct::tagged_pointer<SystemTag, int> int_ptr_end = int_ptr_beg + n;
+  failing_tests += ASSERT_EQUAL(
+      int_ptr_name + " add operator",
+      static_cast<int *>(int_ptr_end) - static_cast<int *>(int_ptr_beg), n);
+
+  dpct::tagged_pointer<SystemTag, int> expect_beg = int_ptr_end - n;
+  failing_tests += ASSERT_EQUAL(int_ptr_name + " subtract operator",
+                                int_ptr_beg == expect_beg, true);
+
+  failing_tests += ASSERT_EQUAL(int_ptr_name + " difference operator",
+                                int_ptr_end - int_ptr_beg, n);
+
+  expect_beg++;
+  expect_beg--;
+  failing_tests +=
+      ASSERT_EQUAL(int_ptr_name + " postfix increment and decrement",
+                   int_ptr_beg == expect_beg, true);
+
+  ++expect_beg;
+  --expect_beg;
+  failing_tests +=
+      ASSERT_EQUAL(int_ptr_name + " prefix increment and decrement",
+                   int_ptr_beg == expect_beg, true);
+
+  expect_beg += 2;
+  expect_beg -= 2;
+  failing_tests +=
+      ASSERT_EQUAL(int_ptr_name + " addition and subtraction assignment",
+                   int_ptr_beg == expect_beg, true);
```

Comment:
This is picky, but this doesn't really check that they do what they should do, only that they are adjoints of each other.

Reply:
True, I have updated this to check the pointer difference after the addition and subtraction assignments (same with the increments / decrements).